File: JavaPractice/src/BinaryTreeFindMinAndMaxValue.java
/***
 * This is for finding min and max value in a given binary tree
 * @author divyesh.shah
 *
 */
public class BinaryTreeFindMinAndMaxValue {

	
	public static class TreeNode {
		int data;
		TreeNode right;
		TreeNode left;

		public TreeNode(int d) {
			this.data = d;
			this.right = null;
			this.left = null;
		}

	}
	public static void main(String arg[]) {

		TreeNode node = new TreeNode(5);
		node.right = new TreeNode(8);
		node.left = new TreeNode(4);
		node.right.left = new TreeNode(13);
		node.left.left = new TreeNode(11);
		node.left.left.right = new TreeNode(2);
		node.right.right = new TreeNode(4);
		node.right.right.right = new TreeNode(1);
		node.left.left.left = new TreeNode(7);
		node.right.right.left = new TreeNode(5);
		node.left.left.left.left = new TreeNode(11);
		node.left.left.left.right = new TreeNode(15);
		node.left.left.left.left.right = new TreeNode(14);
		node.left.left.left.right.right = new TreeNode(16);
		node.left.left.left.right.right.left = new TreeNode(17);
		System.out.println(findMinValueBinaryTree(node));
		System.out.println(findMaxValueBinaryTree(node));

	}

	public static int findMinValueBinaryTree(TreeNode root) {

		if(root == null)
			return Integer.MAX_VALUE;

		int min = root.data;

		int lmin = findMinValueBinaryTree(root.left);
		int rmin = findMinValueBinaryTree(root.right);

		if(lmin < min)
			min = lmin;
		if(rmin < min)
			min = rmin;

		//System.out.println("Min " + min + " Max " + max);
		return min;
	}

	public static int findMaxValueBinaryTree(TreeNode root) {

		if(root == null)
			return Integer.MIN_VALUE;

		int max = root.data;

		int lmax = findMaxValueBinaryTree(root.left);
		int rmax = findMaxValueBinaryTree(root.right);

		if(lmax > max)
			max = lmax;
		if(rmax > max)
			max = rmax;

		//System.out.println("Min " + min + " Max " + max);
		return max;
	}
}

File: PythonPractice/src/BinaryTreeFindMinAndMaxValue.py
from typing import Optional

class TreeNode:
    def __init__(self, data: int):
        if not isinstance(data, int):
            raise TypeError("Data must be an integer")

        self.data: int = data
        self.left: Optional['TreeNode'] = None
        self.right: Optional['TreeNode'] = None

def find_min_value_binary_tree(root: Optional[TreeNode]) -> int:
    if root is None:
        return float('inf')  # Using infinity to represent the maximum possible value

    min_value = root.data

    lmin = find_min_value_binary_tree(root.left)
    rmin = find_min_value_binary_tree(root.right)

    if lmin < min_value:
        min_value = lmin
    if rmin < min_value:
        min_value = rmin

    return min_value

def find_max_value_binary_tree(root: Optional[TreeNode]) -> int:
    if root is None:
        return float('-inf')  # Using negative infinity to represent the minimum possible value

    max_value = root.data

    lmax = find_max_value_binary_tree(root.left)
    rmax = find_max_value_binary_tree(root.right)

    if lmax > max_value:
        max_value = lmax
    if rmax > max_value:
        max_value = rmax

    return max_value

if __name__ == "__main__":
    # Create the binary tree
    node = TreeNode(5)
    node.right = TreeNode(8)
    node.left = TreeNode(4)
    node.right.left = TreeNode(13)
    node.left.left = TreeNode(11)
    node.left.left.right = TreeNode(2)
    node.right.right = TreeNode(4)
    node.right.right.right = TreeNode(1)
    node.left.left.left = TreeNode(7)
    node.right.right.left = TreeNode(5)
    node.left.left.left.left = TreeNode(11)
    node.left.left.left.right = TreeNode(15)
    node.left.left.left.left.right = TreeNode(14)
    node.left.left.left.right.right = TreeNode(16)
    node.left.left.left.right.right.left = TreeNode(17)

    print("Minimum value:", find_min_value_binary_tree(node))
    print("Maximum value:", find_max_value_binary_tree(node))
